AsyncDisplayKit/Details/ASDataController.mm

```diff
@@ -645,240 +524,194 @@ - (void)updateWithChangeSet:(_ASHierarchyChangeSet *)changeSet animated:(BOOL)an
   
   ASDataControllerLogEvent(self, @"triggeredUpdate: %@", changeSet);
   
-  [self beginUpdates];
+  // Step 1: update _sections and _nodeContexts.
+  // After this step, those properties are up-to-date with dataSource's index space.
+  [self _updateSectionContextsWithChangeSet:changeSet];
+  [self _updateNodeContextsWithChangeSet:changeSet];
   
-  for (_ASHierarchyItemChange *change in [changeSet itemChangesOfType:_ASHierarchyChangeTypeDelete]) {
-    [self deleteRowsAtIndexPaths:change.indexPaths withAnimationOptions:change.animationOptions];
-  }
+  // Prepare loadingNodeContexts to be used in editing queue. Deep copy is critical here,
+  // or future edits to the sub-arrays will pollute state between _nodeContexts
+  // and _completedNodeContexts on different threads.
+  ASNodeContextTwoDimensionalDictionary *loadingNodeContexts = [ASDataController deepImmutableCopyOfNodeContextsDictionary:_nodeContexts];
   
-  for (_ASHierarchySectionChange *change in [changeSet sectionChangesOfType:_ASHierarchyChangeTypeDelete]) {
-    [self deleteSections:change.indexSet withAnimationOptions:change.animationOptions];
-  }
-  
-  for (_ASHierarchySectionChange *change in [changeSet sectionChangesOfType:_ASHierarchyChangeTypeInsert]) {
-    [self insertSections:change.indexSet withAnimationOptions:change.animationOptions];
-  }
-  
-  for (_ASHierarchyItemChange *change in [changeSet itemChangesOfType:_ASHierarchyChangeTypeInsert]) {
-    [self insertRowsAtIndexPaths:change.indexPaths withAnimationOptions:change.animationOptions];
-  }
-  
-#if ASEVENTLOG_ENABLE
-  NSString *changeSetDescription = ASObjectDescriptionMakeTiny(changeSet);
-  batchCompletion = ^(BOOL finished) {
-    if (batchCompletion != nil) {
-      batchCompletion(finished);
-    }
-    ASDataControllerLogEvent(self, @"finishedUpdate: %@", changeSetDescription);
-  };
-#endif
-  
-  [self endUpdatesAnimated:animated completion:batchCompletion];
+  dispatch_group_async(_editingTransactionGroup, _editingTransactionQueue, ^{
+    // Step 2: Layout **all** new node contexts without batching in background.
+    // This step doesn't change any internal state.
+    NSArray<ASIndexedNodeContext *> *unloadedContexts = [ASDataController unloadedNodeContextsFromDictionary:loadingNodeContexts];
+    [self batchLayoutNodesFromContexts:unloadedContexts batchSize:unloadedContexts.count batchCompletion:^(id, id) {
+      ASSERT_ON_EDITING_QUEUE;
+      [_mainSerialQueue performBlockOnMainThread:^{
+        // Because loadingNodeContexts is immutable, it can be safely assigned to _loadNodeContexts instead of deep copied.
+        _completedNodeContexts = loadingNodeContexts;
+        
+        // Step 3: Now that _completedNodeContexts is ready, call UIKit to update using the original change set.
+        [self _forwardChangeSetToDelegate:changeSet animated:animated];
```

Comment:
How hard would it be to fuse `reloadData` into `updateWithChangeSet:` by adding a flag like `_ASHierarchyChangeSet.includesReloadData`? Much of the internal processing of `_ASHierarchyChangeSet` would be skipped when that flag is set, and we use it to determine e.g. how to report to the delegate.

That way there's only one implementation of this core `update` method.